File: example/relation_schema.py
from edu.cug.cs.dbp.tutorials.c06.attribute_set import AttributeSet
# from edu.cug.cs.dbp.tutorials.c06.functional_dependency import FunctionalDependency
from edu.cug.cs.dbp.tutorials.c06.functional_dependency_set import FunctionalDependencySet


# 求关系模式rs中，属性集a_set关于F的闭包
def get_closure_of_attribute_set(u_set, fd_set, a_set):
    u = u_set
    fds = fd_set.get_functional_dependency_list().copy()
    if a_set.is_subset(u):
        a = AttributeSet()
        a.copy(a_set)
        b = AttributeSet()
        while True:
            t = AttributeSet()
            del_list = []
            # 依次取函数依赖的左部，判断其左部是否为aset的子集，如果是，则将其右部加入到结果中
            for f in fds:
                las = f.get_left_attribute_set()
                if las.is_subset(a):
                    t.append_all(f.get_right_attribute_group())
                    del_list.append(f)
            # 删除需要删除的函数依赖
            for d in del_list:
                fds.remove(d)
            b = a
            a = a | t
            # 如果等于全集或上一步的集合，则a就是闭包计算的结果
            if a == b or a == u:
                break
        a.sort()
        return a
    else:
        return AttributeSet()


# 求关系模式rs中，属性集a_set关于F的闭包
def get_closure_of_attribute_set_2(rs, a_set):
    return get_closure_of_attribute_set(rs.get_attribute_set(), rs.get_functional_dependency_set(), a_set)


# 求关系模式中F的最小覆盖集
def get_canonical_cover(rs):
    # 1 所有函数依赖的右部分裂为单项属性
    rs.get_functional_dependency_set().right_split()
    # 2 F , G
    u = rs.get_attribute_set()
    fds = rs.get_functional_dependency_set()
    g = fds.clone()
    c = fds.count()
    for i in range(0, c):
        f = fds[i]
        g.remove(f)
        cat = get_closure_of_attribute_set(u, g, f.get_left_attribute_set())
        f_right = f.get_right_attribute_set()
        if f_right.is_subset(cat):
            continue
        else:
            g.append(f)
    # 3 处理左部的冗余属性
    pass
    return g


# 关系模式
class RelationSchema:
    def __init__(self):
        self.fd_set = FunctionalDependencySet()
        self.a_set = AttributeSet()

    def __str__(self):
        s = 'U=' + self.a_set.__str__()
        s += '\n'
        s += 'F=' + self.fd_set.__str__()
        return s

    def get_attribute_set(self):
        return self.a_set

    def get_functional_dependency_set(self):
        return self.fd_set

    def add_attribute(self, attribute):
        self.a_set.append(attribute)

    def add_functional_dependency(self, f):
        self.fd_set.append(f)

File: Closures/Relation_Schema.java
import java.util.*;

public class Relation_Schema
{
    function_dependency_set fd_set;
    Attribute_Set a_set;
    public Relation_Schema(){
        fd_set=new function_dependency_set();
        a_set=new Attribute_Set();

    }
    

    // 转换为字符串
    public String toString(){
        String res="U=";
        res+=this.a_set.toString();
        res+="\n";
        res+="F=";
        res+=this.fd_set.toString();
        return res;
    }

    // 获取属性集
    public Attribute_Set get_Attribute_Set(){
        return this.a_set;
    }
    // 获取函数依赖；
    public function_dependency_set get_Function_dependency_set(){
        return this.fd_set;
    }

    // 添加属性
    public void add_attribute(Attribute a){

        this.a_set.append(a);
    }

    // 添加函数依赖
    public void add_function_dependency(function_dependency a){
        this.fd_set.append(a);
    }

    // 在关系模式rs中，求属性集a_set关于F的闭包
    public  Attribute_Set get_closure_of_attribute_set(function_dependency_set fd_set,Attribute_Set a_set,Attribute_Set u_set){
        ArrayList<function_dependency>fds=new  ArrayList<function_dependency>(fd_set.get_Function_dependency_set());
        // 假如是全集的子集
        if (a_set.is_subset(u_set))
        {
            // 新建两个属性集合
            Attribute_Set a=new Attribute_Set();
            a.copy(a_set);
            Attribute_Set b=new Attribute_Set();
            
            while (true)

            {
                Attribute_Set t= new Attribute_Set();
                // 保存要删除的函数依赖
                ArrayList<function_dependency>dels=new ArrayList<function_dependency>();
                // 依次取函数依赖的左部，判断左部是否为a_set子集
                for (int i=0;i<fds.size();i++){
                    function_dependency f=fds.get(i);
                    Attribute_Set las=f.get_left_Attribute_Set();
                    // 假如是它的子集
                    if (las.is_subset(a)){
                        t.append_list(f.get_right_Attribute_Group());
                        dels.add(f);
                    }
                }
                // 删除要删除的函数依赖
                for(int j=0;j<dels.size();j++){
                    // 删除相应位置的函数依赖
                    fds.remove(dels.get(j));
                }
                b=a.clone();
                // 或运算
                a=a.or_(t);
                if (a.attributes.toString().equals(b.attributes.toString())||a.attributes.toString().equals(u_set.attributes.toString())){
                    break;
                }

            }
            a.sort();
            return a;

        }
        // 不满足直接返回全集
        else{
        return a_set;
        }


    }

    // 求关系模式中F的最小覆盖
    public function_dependency_set get_canonical_cover(Relation_Schema rs){
    //    对右半部分进行分割
        rs.get_Function_dependency_set().right_split();
        // 获取属性全集
        Attribute_Set u=rs.get_Attribute_Set();
        // 获取函数依赖集
        function_dependency_set fds=rs.get_Function_dependency_set();
        // 复制函数依赖
        function_dependency_set g=fds.clone();
        // 获取数目
        int c=fds.count();


        // 按照数目进行迭代
        for(int i=0;i<c;i++){
            // 获取对应位置的函数依赖
            function_dependency f=fds.get_item(i);
            // 删除这个函数依赖
            g.remove(f);
           
            // 计算闭包
            Attribute_Set cat=rs.get_closure_of_attribute_set(g,f.get_left_Attribute_Set(),u);
            
            // 判断子集
            Attribute_Set f_right=f.get_right_Attribute_Set();
            if (f_right.is_subset(cat))continue;
            else{
                g.append(f);
            }
        }

        // 处理左边的冗余属性
        rs.get_Function_dependency_set().left_split();;
        // 获取函数依赖集
        fds=rs.get_Function_dependency_set();
        // 复制函数依赖
        g=fds.clone();
        // 获取数目
        c=fds.count();
        // 按照数目进行迭代
        for(int i=0;i<c;i++){
            // 获取对应位置的函数依赖
            function_dependency f=fds.get_item(i);
            // 删除这个函数依赖
            g.remove(f);
           
            // 计算闭包
            Attribute_Set cat=rs.get_closure_of_attribute_set(g,f.get_left_Attribute_Set(),u);
            
            // 判断子集
            Attribute_Set f_right=f.get_right_Attribute_Set();
            if (f_right.is_subset(cat))continue;
            else{
                g.append(f);
            }
        }

        return g;
    
    }


}
